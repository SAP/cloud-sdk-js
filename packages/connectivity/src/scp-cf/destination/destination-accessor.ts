import { createLogger, ErrorWithCause } from '@sap-cloud-sdk/util';
import { exchangeToken, isTokenExchangeEnabled } from '../identity-service';
import { parseSubdomain } from '../subdomain-replacer';
import {
  DestinationOrFetchOptions,
  sanitizeDestination,
  toDestinationNameUrl
} from './destination';
import { Destination } from './destination-service-types';
import { searchEnvVariablesForDestination } from './destination-from-env';
import {
  DestinationForServiceBindingOptions,
  searchServiceBindingForDestination
} from './destination-from-vcap';
import {
  getDestinationFromDestinationService,
  DestinationFromServiceRetriever,
  getDestinationServiceCredentials
} from './destination-from-service';
import {
  DestinationFetchOptions,
  isDestinationFetchOptions,
  AllDestinationOptions,
  DestinationWithoutToken
} from './destination-accessor-types';
import { searchRegisteredDestination } from './destination-from-registration';
import {
  fetchInstanceDestinations,
  fetchSubaccountDestinations
} from './destination-service';

const logger = createLogger({
  package: 'connectivity',
  messageContext: 'destination-accessor'
});

/**
 * Returns the parameter if it is a destination, calls {@link getDestination} otherwise (which will try to fetch the destination
 * from the Cloud Foundry destination service).
 *
 * Fetching a destination requires:
 * - A binding to exactly one XSUAA service instance with service plan "application".
 * - A binding to a destination service instance.
 *
 * If either of the prerequisites is not met or one of the services returns an error, this function will either throw an error or return a promise that rejects.
 * @param destination - A destination or the necessary parameters to fetch one.
 * @returns A promise resolving to the requested destination on success.
 */
export async function useOrFetchDestination(
  destination: DestinationOrFetchOptions
): Promise<Destination | null> {
  return isDestinationFetchOptions(destination)
    ? getDestination(destination)
    : sanitizeDestination(destination);
}

/**
 * Resolve a destination by the following steps:
 * 1. Call [[useOrFetchDestination]]
 * 2. Throw an error, when the resulting destination from the previous step is falsy
 * 3. Return the checked destination.
 * @param destination - A destination or the necessary parameters to fetch one.
 * @returns A promise resolving to the requested destination on success.
 * @internal
 */
export async function resolveDestination(
  destination: DestinationOrFetchOptions
): Promise<Destination> {
  const resolvedDestination = await useOrFetchDestination(destination).catch(
    error => {
      throw new ErrorWithCause('Failed to load destination.', error);
    }
  );
  if (!resolvedDestination) {
    throw Error(
      `Failed to resolve the destination '${toDestinationNameUrl(
        destination
      )}'.`
    );
  }
  return resolvedDestination;
}

/**
 * Builds a destination from one of three sources (in the given order):
 * - from the environment variable "destinations".
 * - from service bindings.
 * - from the destination service.
 *
 * If you want to get a destination only from a specific source, use the corresponding function directly
 * (`getDestinationFromEnvByName`, `destinationForServiceBinding`, `getDestinationFromDestinationService`).
 * @param options - The options to retrieve the destination.
 * @returns A promise returning the requested destination on success.
 */
export async function getDestination(
  options: DestinationFetchOptions & DestinationForServiceBindingOptions
): Promise<Destination | null> {
  const destination =
    searchEnvVariablesForDestination(options) ||
    (await searchRegisteredDestination(options)) ||
    (await searchServiceBindingForDestination(options)) ||
    (await getDestinationFromDestinationService(options));
  return destination;
}

/**
 * Creates comprehensive log messages from a destinations array and their origin.
 * @param origin - Origin of the destination.
 * @param destinations - Array of destinations.
 * @returns Logs of the retrival of destinations.
 */
function createDestinationFetchLogs(
  origin: string,
  destinations: Destination[]
): string {
  return destinations.reduce(
    (prevLogMessages, currentDestination) =>
      prevLogMessages +
      `Retrieving ${origin} destination: ${currentDestination.name}.\n`,
    ''
  );
}

/**
 * Fetches all destinations from the destination service which match the token.
 * With a subscriber token it fetches all subscriber destinations, otherwise all provider destinations.
 * @param options - The {@link AllDestinationOptions | options} to fetch all destinations.
 * @returns A promise of an array of all destinations without authTokens from the destination service, on success.
 */
export async function getAllDestinationsFromDestinationService(
  options: AllDestinationOptions = {}
): Promise<DestinationWithoutToken[]> {
  logger.debug(
    'Attempting to retrieve all destinations from destination service.'
  );
  if (isTokenExchangeEnabled(options)) {
    options.jwt = await exchangeToken(options);
  }

  const token =
    (await DestinationFromServiceRetriever.getSubscriberToken(options))
      ?.serviceJwt ||
    (await DestinationFromServiceRetriever.getProviderServiceToken(options));

  const destinationServiceUri = getDestinationServiceCredentials().uri;
  const accountName = parseSubdomain(token.decoded.iss!);
  logger.debug(
    `Retrieving all destinations for account: "${accountName}" from destination service.`
  );

  const [instance, subaccount] = await Promise.all([
    fetchInstanceDestinations(destinationServiceUri, token.encoded, options),
    fetchSubaccountDestinations(destinationServiceUri, token.encoded, options)
  ]);

  const loggerMessage =
    createDestinationFetchLogs('instance', instance) +
    createDestinationFetchLogs('subaccount', subaccount);

  logger.debug(loggerMessage);

  const allDestinations = instance.concat(subaccount);

  if (allDestinations && allDestinations.length !== 0) {
    logger.debug(
      `Successfully retrieved all destinations for account: "${accountName}" from destination service.`
    );
  } else {
    logger.debug("Didn't receive any destinations from destination service.");
    return [];
  }

  const allDestinationsWithoutToken: DestinationWithoutToken[] =
    allDestinations.map(destination => {
      delete destination.authTokens;
      return destination;
    });

  return allDestinationsWithoutToken;
}
