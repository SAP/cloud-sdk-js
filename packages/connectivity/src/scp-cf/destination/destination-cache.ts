import { createLogger, first } from '@sap-cloud-sdk/util';
import { Cache, CacheEntry } from '../cache';
import { tenantId } from '../tenant';
import { userId } from '../user';
import { JwtPayload } from '../jsonwebtoken-type';
import { Destination } from './destination-service-types';
import { DestinationsByType } from './destination-accessor-types';

const logger = createLogger({
  package: 'connectivity',
  messageContext: 'destination-cache'
});

/**
 * Enumerator that selects the isolation type of destination in cache.
 * The used isolation strategy is either `Tenant` or `Tenant_User` because we want to get results for subaccount and provider tenants which rules out no-isolation or user isolation.
 */
export enum IsolationStrategy {
  Tenant = 'Tenant',
  Tenant_User = 'TenantUser'
}

/**
 * Interface to implement custom destination caching.
 * To set a custom implementation, call method {@link setDestinationCache} and pass the cache instance.
 */
export interface DestinationCacheInterface {
  hasKey(key: string): Promise<boolean>;
  get(key: string | undefined): Promise<Destination | undefined>;
  set(key: string | undefined, item: CacheEntry<Destination>): Promise<void>;
  clear(): Promise<void>;
}

/**
 * @internal
 * This wrapper class wraps methods of {@link Cache} class as asynchronous methods.
 */
export class DefaultDestinationCache implements DestinationCacheInterface {
  cache: Cache<Destination>;

  constructor(validityTimeInMs?: number) {
    this.cache = new Cache<Destination>(validityTimeInMs);
  }

  /**
   * Specifies whether an entry with a given key is defined in cache.
   * @param key - The entry's key.
   * @returns A boolean value that indicates whether the entry exists in cache.
   */
  async hasKey(key: string): Promise<boolean> {
    return this.cache.hasKey(key);
  }

  /**
   * Getter of cached entries.
   * @param key - The key of the entry to retrieve.
   * @returns The corresponding entry to the provided key if it is still valid, returns `undefined` otherwise.
   */
  async get(key: string | undefined): Promise<Destination | undefined> {
    return this.cache.get(key);
  }

  /**
   * Setter of entries in cache.
   * @param key - The entry's key.
   * @param item - The entry to cache.
   */
  async set(
    key: string | undefined,
    item: CacheEntry<Destination>
  ): Promise<void> {
    return this.cache.set(key, item);
  }

  /**
   * Clear all cached items.
   */
  async clear(): Promise<void> {
    return this.cache.clear();
  }
}

/**
 * @internal
 */
export interface DestinationCacheType {
  /**
   * @internal
   */
  retrieveDestinationFromCache: (
    decodedJwt: Record<string, any>,
    name: string,
    isolation: IsolationStrategy
  ) => Promise<Destination | undefined>;
  /**
   * @internal
   */
  cacheRetrievedDestination: (
    decodedJwt: Record<string, any>,
    destination: Destination,
    isolation: IsolationStrategy
  ) => Promise<void>;
  /**
   * @internal
   */
  cacheRetrievedDestinations: (
    decodedJwt: Record<string, any>,
    retrievedDestinations: DestinationsByType,
    isolation: IsolationStrategy
  ) => Promise<void>;
  /**
   * @internal
   */
  clear: () => Promise<void>;
  /**
   * @internal
   */
  getCacheInstance: () => DestinationCacheInterface;
}

/**
 * DestinationCache constructor.
 * @param cache - Cache object which is used in DestiantionCache
 * @returns A destination cache object.
 * @internal
 */
export const DestinationCache = (
  cache: DestinationCacheInterface = new DefaultDestinationCache(300000)
): DestinationCacheType => ({
  retrieveDestinationFromCache: async (
    decodedJwt: Record<string, any>,
    name: string,
    isolation: IsolationStrategy
  ): Promise<Destination | undefined> =>
    cache.get(getDestinationCacheKey(decodedJwt, name, isolation)),
  cacheRetrievedDestination: async (
    decodedJwt: Record<string, any>,
    destination: Destination,
    isolation: IsolationStrategy
  ): Promise<void> => {
    cacheRetrievedDestination(decodedJwt, destination, isolation, cache);
  },
  cacheRetrievedDestinations: async (
    decodedJwt: Record<string, any>,
    retrievedDestinations: DestinationsByType,
    isolation: IsolationStrategy
  ): Promise<void> => {
    retrievedDestinations.subaccount.forEach(dest =>
      cacheRetrievedDestination(decodedJwt, dest, isolation, cache)
    );
    retrievedDestinations.instance.forEach(dest =>
      cacheRetrievedDestination(decodedJwt, dest, isolation, cache)
    );
  },
  clear: async (): Promise<void> => {
    cache.clear();
  },
  getCacheInstance: () => cache
});

/**
 * Calculates a cache key based on the jwt and destination name for the given isolation strategy.
 * Cache keys for strategies are non-overlapping, i.e. using a cache key for strategy {@link IsolationStrategy.Tenant}
 * will not result in a cache hit for a destination that has been cached with strategy {@link IsolationStrategy.Tenant_User}.
 * @param decodedJwt - The decoded JWT of the current request.
 * @param destinationName - The name of the destination.
 * @param isolationStrategy - The strategy used to isolate cache entries.
 * @returns The cache key.
 * @internal
 */
export function getDestinationCacheKey(
  decodedJwt: Record<string, any>,
  destinationName: string,
  isolationStrategy = IsolationStrategy.Tenant_User
): string | undefined {
  const tenant = tenantId(decodedJwt);
  const user = userId(decodedJwt);
  switch (isolationStrategy) {
    case IsolationStrategy.Tenant:
      if (tenant) {
        return `${tenant}::${destinationName}`;
      }
      logger.warn(
        `Cannot get cache key. Isolation strategy ${isolationStrategy} is used, but tenant id is undefined.`
      );
      return;
    case IsolationStrategy.Tenant_User:
      if (tenant && user) {
        return `${user}:${tenant}:${destinationName}`;
      }
      logger.warn(
        `Cannot get cache key. Isolation strategy ${isolationStrategy} is used, but tenant id or user id is undefined.`
      );
      return;
    default:
      logger.warn(
        `Cannot get cache key. Isolation strategy ${isolationStrategy} is not supported.`
      );
      return;
  }
}

async function cacheRetrievedDestination<T extends DestinationCacheInterface>(
  decodedJwt: Record<string, any>,
  destination: Destination,
  isolation: IsolationStrategy,
  cache: T
): Promise<void> {
  if (!destination.name) {
    throw new Error('The destination name is undefined.');
  }

  const key = getDestinationCacheKey(decodedJwt, destination.name, isolation);
  const expiresIn = first(destination.authTokens || [])?.expiresIn;
  const expirationTime = expiresIn
    ? Date.now() + parseInt(expiresIn) * 1000
    : undefined;
  cache.set(key, { entry: destination, expires: expirationTime });
}

/**
 * Sets the custom destination cache instance.
 * Call this method with an instance of {@link DestinationCacheInterface} to override the default cache instance set by the SDK.
 *
 * NOTE: This function should be called at the beginning before any calls to either {@link getDestination} or {@link @sap-cloud-sdk/http-client!executeHttpRequest}.
 * @param cache - An instance of {@link DestinationCacheInterface}.
 */
export function setDestinationCache(cache: DestinationCacheInterface): void {
  destinationCache = DestinationCache(cache);
}

/**
 * @internal
 */
export let destinationCache: DestinationCacheType = DestinationCache();
/**
 * Determin the default Isolation strategy if not given as option.
 * @param jwt - JWT to determine the default isolation strategy
 * @returns The isolation strategy based on the JWT. If no JWT is given it defaults to Tenant isolation
 * @internal
 */
export function getDefaultIsolationStrategy(
  jwt: JwtPayload | undefined
): IsolationStrategy {
  return jwt && userId(jwt)
    ? IsolationStrategy.Tenant_User
    : IsolationStrategy.Tenant;
}
