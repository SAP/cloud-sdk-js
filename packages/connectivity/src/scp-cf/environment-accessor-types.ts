/**
 * Unspecific representation of a service as read from VCAP_SERVICES (for Cloud Foundry) or mounted secrets (for K8S).
 */
export interface Service {
  [other: string]: any;
  /**
   * The name of the service.
   */
  name: string;
  /**
   * The label of the service.
   */
  label: string;
  /**
   * The tags of the service.
   */
  tags: string[];
  /**
   * The service credentials.
   */
  credentials: ServiceCredentials;
}

/**
 * Unspecific representation of a service's credentials as read from VCAP_SERVICES (for Cloud Foundry) or mounted secrets (for K8S).
 */
export type ServiceCredentials = {
  [other: string]: any;
  clientid: string;
} & (
  | {
      clientsecret: string;
    }
  | {
      certificate: string;
      key: string;
    }
);

/**
 * Credentials for the Destination service.
 * @internal
 */
export type DestinationServiceCredentials = ServiceCredentials & {
  identityzone: string;
  instanceid: string;
  tenantid: string;
  tenantmode: string;
  uaadomain: string;
  uri: string;
  url: string;
  verificationkey: string;
  xsappname: string;
};

/**
 * Credentials for the XSUAA service.
 */
export type XsuaaServiceCredentials = ServiceCredentials & {
  identityzone: string;
  identityzoneid: string;
  sburl: string;
  tenantid: string;
  tenantmode: string;
  uaadomain: string;
  url: string;
  verificationkey: string;
  xsappname: string;
};
