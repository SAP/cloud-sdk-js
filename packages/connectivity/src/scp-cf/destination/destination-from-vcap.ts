import { createLogger } from '@sap-cloud-sdk/util';
import { JwtPayload } from '../jsonwebtoken-type';
import { decodeJwt } from '../jwt';
import { Service } from '../environment-accessor-types';
import { getServiceByInstanceName } from '../environment-accessor';
import {
  addProxyConfigurationInternet,
  ProxyStrategy,
  proxyStrategy
} from './http-proxy-util';
import { Destination } from './destination-service-types';
import type { DestinationFetchOptions } from './destination-accessor-types';
import { destinationCache, IsolationStrategy } from './destination-cache';
import { decodedJwtOrZid } from './destination-from-registration';
import { serviceToDestinationTransformers } from './service-binding-to-destination';

const logger = createLogger({
  package: 'connectivity',
  messageContext: 'destination-accessor-vcap'
});

/**
 * Represents partial options to fetch destinations.
 */
export interface PartialDestinationFetchOptions {
  /**
   * The fetched destination will be cached if set to true.
   */
  useCache?: boolean;
  /**
   * The jwt payload used to fetch destinations.
   */
  jwt?: JwtPayload;
}

/**
 * Tries to build a destination from a service binding with the given name.
 * Throws an error if no services are bound at all, no service with the given name can be found, or the service type is not supported.
 * The last error can be circumvent by using the second parameter to provide a custom function that transforms a service binding to a destination.
 * @param serviceInstanceName - The name of the service.
 * @param options - Options to customize the behavior of this function.
 * @returns A destination.
 */
export async function destinationForServiceBinding(
  serviceInstanceName: string,
  options: DestinationForServiceBindingOptions &
    PartialDestinationFetchOptions = {}
): Promise<Destination> {
  if (options.useCache) {
    const fromCache = await destinationCache.retrieveDestinationFromCache(
      options.jwt || decodedJwtOrZid().subaccountid,
      serviceInstanceName,
      IsolationStrategy.Tenant
    );
    if (fromCache) {
      return fromCache;
    }
  }

  const selected = getServiceByInstanceName(serviceInstanceName)!;
  const destination = options.serviceBindingTransformFn
    ? await options.serviceBindingTransformFn(selected, options)
    : await transform(selected, options);

  const destWithProxy =
    destination &&
    (proxyStrategy(destination) === ProxyStrategy.INTERNET_PROXY ||
      proxyStrategy(destination) === ProxyStrategy.PRIVATELINK_PROXY)
      ? addProxyConfigurationInternet(destination)
      : destination;

  if (options.useCache) {
    // use the provider tenant if no jwt is given. Since the grant type is clientCredential isolation strategy is tenant.
    await destinationCache.cacheRetrievedDestination(
      options.jwt || decodedJwtOrZid().subaccountid,
      destWithProxy,
      IsolationStrategy.Tenant
    );
  }

  return destWithProxy;
}

/**
 * Options to customize the behavior of {@link destinationForServiceBinding}.
 */
export interface DestinationForServiceBindingOptions {
  /**
   * Custom transformation function to control how a {@link Destination} is built from the given {@link Service}.
   */
  serviceBindingTransformFn?: ServiceBindingTransformFunction;
}

/**
 * Type of the function to transform the service binding.
 */
export type ServiceBindingTransformFunction = (
  service: Service,
  options?: PartialDestinationFetchOptions
) => Promise<Destination>;

async function transform(
  service: Service,
  options?: PartialDestinationFetchOptions
): Promise<Destination> {
  if (!serviceToDestinationTransformers[service.label]) {
    throw serviceTypeNotSupportedError(service);
  }

  return serviceToDestinationTransformers[service.label](service, options);
}

function serviceTypeNotSupportedError(service: Service): Error {
  return Error(`The service "${service.name}" is of type "${service.label}" which is not supported! Consider providing your own transformation function when calling destinationForServiceBinding, like this:
  destinationServiceForBinding(yourServiceName, { serviceBindingToDestination: yourTransformationFunction });`);
}

/**
 * @internal
 */
export async function searchServiceBindingForDestination({
  iss,
  jwt,
  serviceBindingTransformFn,
  destinationName,
  useCache
}: DestinationFetchOptions &
  DestinationForServiceBindingOptions): Promise<Destination | null> {
  logger.debug('Attempting to retrieve destination from service binding.');
  try {
    const jwtFromOptions = iss ? { iss } : jwt ? decodeJwt(jwt) : undefined;
    const useCacheIgnoringUndefined =
      typeof useCache !== 'undefined' ? { useCache } : {};
    const destination = await destinationForServiceBinding(destinationName, {
      serviceBindingTransformFn,
      jwt: jwtFromOptions,
      ...useCacheIgnoringUndefined
    });
    logger.info('Successfully retrieved destination from service binding.');
    return destination;
  } catch (error) {
    logger.debug(
      `Could not retrieve destination from service binding. If you are not using SAP Extension Factory, this information probably does not concern you. ${error.message}`
    );
  }
  return null;
}
