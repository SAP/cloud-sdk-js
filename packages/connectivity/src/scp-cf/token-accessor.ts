import { ErrorWithCause } from '@sap-cloud-sdk/util';
import { JwtPayload } from './jsonwebtoken-type';
import { decodeJwt } from './jwt';
import { CachingOptions } from './cache';
import { clientCredentialsTokenCache } from './client-credentials-token-cache';
import {
  getXsuaaServiceCredentials,
  resolveService
} from './environment-accessor';
import { Service, XsuaaServiceCredentials } from './environment-accessor-types';
import { ResilienceOptions } from './resilience-options';
import { replaceSubdomain } from './subdomain-replacer';
import { getClientCredentialsToken, getUserToken } from './xsuaa-service';

/**
 * Returns an access token that can be used to call the given service. The token is fetched via a client credentials grant with the credentials of the given service.
 * If multiple instances of the provided service exist, the first instance will be selected.
 * When a JWT is passed, the tenant of the JWT will be used when performing the grant.
 * When no JWT is passed, the grant will be performed using the provider tenant.
 *
 * Throws an error if there is no instance of the given service type or the XSUAA service, or if the request to the XSUAA service fails.
 * @param service - The type of the service or an instance of {@link Service}.
 * @param options - Options to influence caching and resilience behavior (see {@link CachingOptions} and {@link ResilienceOptions}, respectively) and a JWT. By default, caching and usage of a circuit breaker are enabled.
 * @returns Access token.
 */
export async function serviceToken(
  service: string | Service,
  options?: CachingOptions &
    ResilienceOptions & {
      jwt?: string | JwtPayload;
      // TODO 2.0 Once the xssec supports caching remove all xsuaa related content here and use their cache.
      xsuaaCredentials?: XsuaaServiceCredentials;
    }
): Promise<string> {
  const opts = {
    useCache: true,
    enableCircuitBreaker: true,
    ...options
  };

  service = resolveService(service);
  const serviceCredentials = service.credentials;

  // TODO 2.0 Once the xssec supports caching remove all xsuaa related content here and use their cache.
  if (opts.useCache) {
    const xsuaa = multiTenantXsuaaCredentials(options);

    const cachedToken = clientCredentialsTokenCache.getToken(
      xsuaa.url,
      serviceCredentials.clientid
    );
    if (cachedToken) {
      return cachedToken.access_token;
    }
  }

  try {
    const token = await getClientCredentialsToken(
      service,
      options?.jwt,
      options
    );

    if (opts.useCache) {
      const xsuaa = multiTenantXsuaaCredentials(options);

      clientCredentialsTokenCache.cacheToken(
        xsuaa.url,
        serviceCredentials.clientid,
        token
      );
    }

    return token.access_token;
  } catch (err) {
    throw new ErrorWithCause(
      `Could not fetch client credentials token for service of type "${service.label}".`,
      err
    );
  }
}

/**
 * Returns a jwt bearer token that can be used to call the given service.
 * The token is fetched via a JWT bearer token grant using the user token + client credentials.
 *
 * Throws an error if there is no instance of the given service type or the XSUAA service, or if the request to the XSUAA service fails.
 * @param jwt - The JWT of the user for whom the access token should be fetched.
 * @param service - The type of the service or an instance of {@link Service}.
 * @param options - Options to influence resilience behavior (see {@link ResilienceOptions}). By default, usage of a circuit breaker is enabled.
 * @returns A jwt bearer token.
 */
export async function jwtBearerToken(
  jwt: string,
  service: string | Service,
  options?: ResilienceOptions
): Promise<string> {
  const resolvedService = resolveService(service);

  const opts: ResilienceOptions = {
    enableCircuitBreaker: true,
    ...options
  };

  return getUserToken(resolvedService, jwt, opts);
}

function multiTenantXsuaaCredentials(
  options: {
    jwt?: string | JwtPayload;
    xsuaaCredentials?: XsuaaServiceCredentials;
  } = {}
): XsuaaServiceCredentials {
  const xsuaa = options.xsuaaCredentials
    ? { ...options.xsuaaCredentials }
    : getXsuaaServiceCredentials(options.jwt);

  if (options.jwt) {
    const decodedJwt =
      typeof options.jwt === 'string' ? decodeJwt(options.jwt) : options.jwt;

    if (!decodedJwt.iss) {
      throw Error('Property `iss` is missing in the provided user token.');
    }

    xsuaa.url = replaceSubdomain(decodedJwt.iss, xsuaa.url);
  }

  return xsuaa;
}
