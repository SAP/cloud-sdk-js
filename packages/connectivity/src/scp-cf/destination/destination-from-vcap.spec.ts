import {
  mockServiceToken,
  providerUserPayload
} from '../../../../../test-resources/test/test-util';
import * as tokenAccessor from '../token-accessor';
import { Service } from '../environment-accessor-types';
import { getDestination } from './destination-accessor';
import { destinationForServiceBinding } from './destination-from-vcap';
import { destinationCache } from './destination-cache';
import SpyInstance = jest.SpyInstance;

describe('vcap-service-destination', () => {
  const spy = jest.spyOn(tokenAccessor, 'serviceToken');

  beforeAll(() => {
    mockServiceToken();
  });

  beforeEach(() => {
    mockServiceBindings();
  });

  afterEach(() => {
    delete process.env.VCAP_SERVICES;
    jest.clearAllMocks();
  });

  function getActualClientId(spyInstance: SpyInstance): string {
    return spyInstance.mock.calls[0][0]['credentials']['clientid'];
  }

  it('creates a destination for the business logging service', async () => {
    await expect(
      destinationForServiceBinding('my-business-logging', {
        jwt: providerUserPayload
      })
    ).resolves.toEqual({
      url: 'https://business-logging.my.example.com',
      authentication: 'OAuth2ClientCredentials',
      name: 'my-business-logging',
      authTokens: [expect.objectContaining({ value: expect.any(String) })]
    });

    expect(getActualClientId(spy)).toBe('clientIdBusinessLogging');
  });

  it('creates ad destination for the xsuaa service', async () => {
    await expect(
      destinationForServiceBinding('my-xsuaa', {
        jwt: providerUserPayload
      })
    ).resolves.toEqual({
      url: 'https://api.authentication.sap.hana.ondemand.com',
      authentication: 'OAuth2ClientCredentials',
      name: 'my-xsuaa',
      authTokens: [expect.objectContaining({ value: expect.any(String) })]
    });

    expect(getActualClientId(spy)).toBe('clientIdXsUaa');
  });

  it('creates a destination for the service manager service', async () => {
    await expect(
      destinationForServiceBinding('my-service-manager', {
        jwt: providerUserPayload
      })
    ).resolves.toEqual({
      url: 'https://service-manager.cfapps.sap.hana.ondemand.com',
      authentication: 'OAuth2ClientCredentials',
      name: 'my-service-manager',
      authTokens: [expect.objectContaining({ value: expect.any(String) })]
    });

    expect(getActualClientId(spy)).toBe('clientIdServiceManager');
  });

  it('creates a destination for the destination service', async () => {
    await expect(
      destinationForServiceBinding('my-destination-service', {
        jwt: providerUserPayload
      })
    ).resolves.toEqual({
      url: 'https://destination-configuration.cfapps.sap.hana.ondemand.com',
      authentication: 'OAuth2ClientCredentials',
      name: 'my-destination-service',
      authTokens: [expect.objectContaining({ value: expect.any(String) })]
    });
    expect(getActualClientId(spy)).toBe('clientIdDestination');
  });

  it('creates a destination for the saas registry', async () => {
    await expect(
      destinationForServiceBinding('my-saas-registry', {
        jwt: providerUserPayload
      })
    ).resolves.toEqual({
      url: 'https://saas-manager.mesh.cf.sap.hana.ondemand.com',
      authentication: 'OAuth2ClientCredentials',
      name: 'my-saas-registry',
      authTokens: [expect.objectContaining({ value: expect.any(String) })]
    });
    expect(getActualClientId(spy)).toBe('clientIdSaasRegistry');
  });

  it('creates a destination for the workflow', async () => {
    await expect(
      destinationForServiceBinding('my-workflow', {
        jwt: providerUserPayload
      })
    ).resolves.toEqual({
      url: 'https://api.workflow-sap.cfapps.sap.hana.ondemand.com/workflow-service/odata',
      authentication: 'OAuth2ClientCredentials',
      name: 'my-workflow',
      authTokens: [expect.objectContaining({ value: expect.any(String) })]
    });
    expect(getActualClientId(spy)).toBe('clientIdWorkFlow');
  });

  it('creates a destination for the XF s4 hana cloud service', async () => {
    await expect(
      destinationForServiceBinding('my-s4-hana-cloud')
    ).resolves.toEqual({
      url: 'https://my.example.com',
      authentication: 'BasicAuthentication',
      username: 'USER_NAME',
      password: 'PASSWORD'
    });
  });

  it('uses the cache if enabled', async () => {
    await destinationCache.clear();

    await destinationForServiceBinding('my-destination-service', {
      useCache: true,
      jwt: providerUserPayload
    });
    await destinationForServiceBinding('my-destination-service', {
      useCache: true,
      jwt: providerUserPayload
    });

    expect(spy).toBeCalledTimes(1);
    expect(
      destinationCache
        .getCacheInstance()
        .get(`${providerUserPayload.zid}::my-destination`)
    ).toBeDefined();
  });

  it('creates a destination using a custom transformation function', async () => {
    const serviceBindingTransformFn = jest.fn(async (service: Service) => ({
      url: service.credentials.sys
    }));

    await expect(
      destinationForServiceBinding('my-custom-service', {
        serviceBindingTransformFn
      })
    ).resolves.toEqual({
      url: 'https://custom-service.my.example.com'
    });
    expect(serviceBindingTransformFn).toBeCalledTimes(1);
  });

  it('throws an error if the service type is not supported', async () => {
    await expect(() =>
      destinationForServiceBinding('my-custom-service')
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('throws an error if no service binding can be found for the given name', async () => {
    await expect(() =>
      destinationForServiceBinding('non-existent-service')
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      '"No service with the name: \\"non-existent-service\\" was found."'
    );
  });

  it('finds the destination when searching for service bindings', async () => {
    const serviceBindingTransformFn = jest.fn(async (service: Service) => ({
      url: service.credentials.sys
    }));

    await expect(
      getDestination({
        destinationName: 'my-custom-service',
        serviceBindingTransformFn
      })
    ).resolves.toEqual({
      url: 'https://custom-service.my.example.com'
    });
    expect(serviceBindingTransformFn).toBeCalledTimes(1);
  });
});

function mockServiceBindings() {
  process.env.VCAP_SERVICES = JSON.stringify(serviceBindings);
}

const serviceBindings = {
  'business-logging': [
    {
      name: 'my-business-logging',
      label: 'business-logging',
      tags: [
        'business-logging',
        'logging',
        'com.sap.appbasic.businesslogs',
        'comsapappbasicbusinesslogs'
      ],
      credentials: {
        writeUrl: 'https://business-logging.my.example.com',
        uaa: {
          clientid: 'clientIdBusinessLogging',
          clientsecret: 'PASSWORD'
        }
      }
    }
  ],
  workflow: [
    {
      name: 'my-workflow',
      label: 'workflow',
      tags: [],
      credentials: {
        endpoints: {
          workflow_odata_url:
            'https://api.workflow-sap.cfapps.sap.hana.ondemand.com/workflow-service/odata',
          workflow_rest_url:
            'https://api.workflow-sap.cfapps.sap.hana.ondemand.com/workflow-service/rest'
        },
        uaa: {
          clientid: 'clientIdWorkFlow',
          clientsecret: 'PASSWORD'
        }
      }
    }
  ],
  destination: [
    {
      name: 'my-destination-service',
      label: 'destination',
      credentials: {
        clientid: 'clientIdDestination',
        clientsecret: 'PASSWORD',
        uri: 'https://destination-configuration.cfapps.sap.hana.ondemand.com'
      }
    }
  ],
  'custom-service': [
    {
      credentials: {
        sys: 'https://custom-service.my.example.com'
      },
      name: 'my-custom-service'
    }
  ],
  's4-hana-cloud': [
    {
      name: 'my-s4-hana-cloud',
      label: 's4-hana-cloud',
      credentials: {
        Password: 'PASSWORD',
        URL: 'https://my.example.com',
        User: 'USER_NAME'
      }
    }
  ],
  'saas-registry': [
    {
      label: 'saas-registry',
      name: 'my-saas-registry',
      credentials: {
        clientid: 'clientIdSaasRegistry',
        clientsecret: 'PASSWORD',
        saas_registry_url: 'https://saas-manager.mesh.cf.sap.hana.ondemand.com'
      }
    }
  ],
  'service-manager': [
    {
      label: 'service-manager',
      name: 'my-service-manager',
      credentials: {
        sm_url: 'https://service-manager.cfapps.sap.hana.ondemand.com',
        clientid: 'clientIdServiceManager',
        clientsecret: 'PASSWORD'
      }
    }
  ],
  xsuaa: [
    {
      label: 'xsuaa',
      name: 'my-xsuaa',
      tags: ['xsuaa'],
      credentials: {
        clientid: 'clientIdXsUaa',
        clientsecret: 'PASSWORD',
        apiurl: 'https://api.authentication.sap.hana.ondemand.com'
      }
    }
  ]
};
