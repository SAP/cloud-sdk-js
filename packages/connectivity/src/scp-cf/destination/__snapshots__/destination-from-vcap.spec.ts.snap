// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vcap-service-destination throws an error if the service type is not supported 1`] = `
"The service \\"my-custom-service\\" is of type \\"undefined\\" which is not supported! Consider providing your own transformation function when calling destinationForServiceBinding, like this:
  destinationServiceForBinding(yourServiceName, { serviceBindingToDestination: yourTransformationFunction });"
`;
